fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java

```diff
@@ -0,0 +1,611 @@
+                // collect dict expression
+                List<ScalarOperator> exprs = dictExpressionCollector.getDictExpressions(c);
+                if (!exprs.isEmpty()) {
+                    // maybe not new dict, just optimize the expression with dictionary
+                    stringExpressions.computeIfAbsent(c, l -> Lists.newArrayList()).addAll(exprs);
+                }
+
+                // whole expression support dictionary, define new dict column, only support varchar column
+                if (exprs.contains(value) && value.getType().isVarchar()) {
+                    stringRefToDefineExprMap.put(key.getId(), value);
+                    expressionStringRefCounter.putIfAbsent(key.getId(), 0);
+                    info.outputStringColumns.union(key.getId());
+                }
+            });
+        }
+    }
+
+    // Check if an expression can be optimized using a dictionary
+    // If the expression only contains a string column, the expression can be optimized using a dictionary
+    private static class DictExpressionCollector extends ScalarOperatorVisitor<ScalarOperator, Void> {
+        private static final ScalarOperator CONSTANTS = ConstantOperator.TRUE;
+        private static final ScalarOperator VARIABLES = ConstantOperator.FALSE;
+
+        private final ColumnRefSet allDictColumnRefs;
+        private final Map<Integer, List<ScalarOperator>> dictExpressions = Maps.newHashMap();
+
+        public DictExpressionCollector(ColumnRefSet allDictColumnRefs) {
+            this.allDictColumnRefs = allDictColumnRefs;
+        }
+
+        public void collect(ScalarOperator scalarOperator) {
+            ScalarOperator res = scalarOperator.accept(this, null);
+            if (res.isColumnRef()) {
+                dictExpressions.computeIfAbsent(((ColumnRefOperator) res).getId(), x -> Lists.newArrayList())
+                        .add(scalarOperator);
+            }
+        }
+
+        public List<ScalarOperator> getDictExpressions(int columnId) {
+            if (!dictExpressions.containsKey(columnId)) {
+                return Collections.emptyList();
+            }
+
+            return dictExpressions.get(columnId);
+        }
+
+        public List<ScalarOperator> visitChildren(ScalarOperator operator, Void context) {
+            List<ScalarOperator> children = Lists.newArrayList();
+            for (ScalarOperator child : operator.getChildren()) {
+                children.add(child.accept(this, context));
+            }
+            return children;
+        }
+
+        private ScalarOperator merge(List<ScalarOperator> collectors, ScalarOperator scalarOperator) {
+            // all constant
+            if (collectors.stream().allMatch(CONSTANTS::equals)) {
+                return CONSTANTS;
+            }
+
+            long variableExpr = collectors.stream().filter(VARIABLES::equals).count();
+            long dictCount = collectors.stream().filter(ScalarOperator::isColumnRef).distinct().count();
```

Comment:
distinct may not works when collectors is nort sorted

Reply:
distinct don't need sorted